sklearn/ensemble/gradient_boosting.py

```diff
@@ -1484,7 +1484,7 @@ def fit(self, X, y, sample_weight=None, monitor=None):
             else:
                 try:
                     self.init_.fit(X, y, sample_weight=sample_weight)
-                except TypeError:
+                except (TypeError, ValueError):
```

Comment:
I feel like here it should be actually more strict catching than loose. A lot of (if not most of) init param validation happens in fit, and they raise a `ValueError` if the parameters are not valid, and here the user would instead see a message complaining about `sample_weights` which would be irrelevant.

Wouldn't checking the signature, or the [appropriate] estimator tag be a better idea here?

Reply:
>the user would instead see a message complaining about sample_weights which would be irrelevant.

Why is it be irrelevant? This is precisely what this check is about.

>Wouldn't checking the signature, or the [appropriate] estimator tag be a better idea here? 

Yes agree, but apparently using a try catch is preferred https://github.com/scikit-learn/scikit-learn/pull/12983#discussion_r254671832

Reply:
It's not elegant but I'm okay with this.

Reply:
For example, NuSVR supports `sample_weights` on fit:

```
X, y = make_regression(random_state=0)
NuSVR(nu=1.5).fit(X, y, sample_weight=np.ones(X.shape[0]))
```
gives:

```
ValueError: nu <= 0 or nu > 1
```

But after this PR:

```
init = make_pipeline(NuSVR(nu=1.5))
gb = GradientBoostingRegressor(init=init)
gb.fit(X, y, sample_weight=np.ones(X.shape[0]))
```
gives:

```
ValueError: The initial estimator Pipeline does not support sample weights.
```

Reply:
Ooh you were talking about the input checking of the init estimator, ok, good point.

@jnothman do you think using `has_fit_param()` would be justified here? As far as I understand, this would only be a problem if a user passes a custom estimator which accepts `sample_weight` in `fit()` as a keyword args.

Another option would be to test the error message of the `ValueError` coming from a pipeline and only raise in this case?

Reply:
I don't see how has_fit_param helps here.

But perhaps we should raise a more equivocal error message ("could not fit init estimator with sample_weight") and use `raise from` to report the original exception

Reply:
>I don't see how has_fit_param helps here.

This check is only supposed to check whether the init estimator supports samples_weights. I had to add `ValueError` for pipelines because unlike traditional estimators, they don't raise `TypeError`. As @adrinjalali noted, now the check also catches ValueError coming from other reasons (namely input checking).

Using `has_fit_param` would avoid this, I think.

Reply:
> Using has_fit_param would avoid this, I think.

How so?

Reply:
Using has_fit_param avoids using a try except